claat/serve.go

```diff
@@ -0,0 +1,221 @@
+	if spec == "^0.7.2" {
+		spec = "webcomponents/webcomponentsjs#^0.7.2"
+	}
+	var user, repo, path, vers string
+	s := strings.Split(spec, "#")
+	path = s[0]
+	if len(s) > 1 {
+		vers = s[1]
+		if s[1][0] == '^' {
+			vers = s[1][1:]
+		}
+	} else {
+		vers = "master"
+	}
+	s = strings.Split(path, "/")
+	user = s[0]
+	if len(s) > 1 {
+		repo = s[1]
+	}
+
+	// if repo already exists locally, return immediately, we're done.
+	if _, err := os.Stat(depsDir + "/" + repo); err == nil {
+		return nil
+	}
+	zipFile := depsDir + "/" + repo + ".zip"
+	url := "https://github.com/" + user + "/" + repo + "/archive/v" + vers + ".zip"
+	err := downloadFile(zipFile, url)
+	if err != nil {
+		return err
+	}
+	// If get fails, it will download a file containing only "404: Not Found".
+	// We check for that case by looking for an unusally small file.
+	var st os.FileInfo
+	if st, err = os.Stat(zipFile); err != nil {
+		return err
+	}
+	if st.Size() < 20 {
+		os.Remove(zipFile)
+		url = "https://github.com/" + user + "/" + repo + "/archive/" + vers + ".zip"
+		err = downloadFile(zipFile, url)
+		if err != nil {
+			return err
+		}
+	}
+	err = unzip(zipFile, depsDir)
+	if err != nil {
+		return err
+	}
+	os.Remove(zipFile)
+	os.Rename(depsDir+"/"+repo+"-"+vers, depsDir+"/"+repo)
+
+	// if unzipped archive contains a bower.json, parse it, and for each dependency therein,
+	// recursively fetch the corresponding repo.
+	bowerFile := depsDir + "/" + repo + "/bower.json"
+	if _, err := os.Stat(bowerFile); os.IsNotExist(err) {
+		return nil
+	}
+	raw, err := ioutil.ReadFile(bowerFile)
+	if err != nil {
+		return err
+	}
+	var b struct {
+		Dependencies map[string]string
+	}
+	err = json.Unmarshal(raw, &b)
+	if err != nil {
+		return err
+	}
+	for _, v := range b.Dependencies {
+		err = fetchRepo(depsDir, v)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// unzip a file to a dest directory.
+func unzip(src, dest string) error {
+	r, err := zip.OpenReader(src)
+	if err != nil {
+		return err
+	}
+	defer r.Close()
+	os.MkdirAll(dest, 0755)
+	// Closure to address file descriptors issue with all the deferred .Close() methods
+	extractAndWriteFile := func(f *zip.File) error {
+		rc, err := f.Open()
+		if err != nil {
+			return err
+		}
+		defer rc.Close()
+		path := filepath.Join(dest, f.Name)
+		if f.FileInfo().IsDir() {
+			return os.MkdirAll(path, f.Mode())
+		}
+		os.MkdirAll(filepath.Dir(path), f.Mode())
+		f2, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, f.Mode())
+		if err != nil {
+			return err
+		}
+		_, err = io.Copy(f2, rc)
+		// Catch both io.Copy and file.Close errors but prefer returning the former.
+		if err1 := f2.Close(); err1 != nil && err == nil {
+			err = err1
+		}
+		return err
+	}
+	for _, f := range r.File {
+		err := extractAndWriteFile(f)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// openBrowser tries to open the URL in a browser.
+func openBrowser(url string) bool {
```

Comment:
what does `bool` signifies, that it opened a url? but the cmd doesn't actually know it. it just executes something.
I'd rather return `error` or nothing at all.

Reply:
Fixed.